File: snobot_sim/cpp/lib/SnobotSim/SimulatorComponents/navx/NavxSimulator.cpp
/*
 * NavxSimulator.cpp
 *
 *  Created on: Jul 8, 2017
 *      Author: PJ
 */

#include "SnobotSim/SimulatorComponents/navx/NavxSimulator.h"
#include "SnobotSim/SensorActuatorRegistry.h"

NavxSimulator::NavxSimulator(int aPort) :
	mYawGyro(new GyroWrapper("NavX Yaw")),
	mPitchGyro(new GyroWrapper("NavX Pitch")),
	mRollGyro(new GyroWrapper("NavX Roll"))
{
	SensorActuatorRegistry::Get().Register((aPort + 1) * 150 + 1, mYawGyro);
	SensorActuatorRegistry::Get().Register((aPort + 1) * 150 + 2, mPitchGyro);
	SensorActuatorRegistry::Get().Register((aPort + 1) * 150 + 3, mRollGyro);
}

NavxSimulator::~NavxSimulator()
{

}

void NavxSimulator::GetWriteConfig(uint8_t* aBuffer)
{
	aBuffer[ 0] = 0x01; // Type
	aBuffer[ 1] = 0x02; // HW rev
	aBuffer[ 2] = 0x03; // FW Major
	aBuffer[ 3] = 0x04; // FW Minor
	aBuffer[ 4] = 0x05; // Update Rate
	aBuffer[ 5] = 0x06; // Accel g
	aBuffer[ 6] = 0x07; // Gyro DPS
	aBuffer[ 7] = 0x08; // Gyro DPS
	aBuffer[ 8] = 0x09; // Op Status
	aBuffer[ 9] = 0x10; // Cal Status
	aBuffer[10] = 0x11; // Self Test Status
	aBuffer[11] = 0x12; // Capability Flags
	aBuffer[12] = 0x13; // Capability Flags
	aBuffer[16] = 0x14; // Sensor Status
	aBuffer[17] = 0x15; // Sensor Status
}

void NavxSimulator::GetCurrentData(uint8_t* aBuffer, int aFirstAddress)
{
	static uint32_t timestamp = 1;

	PutTheValue(aBuffer, 0x12 - aFirstAddress, timestamp, 4); // Op Status
	++timestamp;

	PutTheValue(aBuffer, 0x08 - aFirstAddress, 1, 1); // Op Status
	PutTheValue(aBuffer, 0x0A - aFirstAddress, 2, 1); // Self Test
	PutTheValue(aBuffer, 0x09 - aFirstAddress, 3, 1); // Cal Status
	PutTheValue(aBuffer, 0x10 - aFirstAddress, 4, 1); // Sensor Status

	PutTheValue(aBuffer, 0x16 - aFirstAddress, int16_t(mYawGyro->GetAngle() * 100), 2); // Yaw
	PutTheValue(aBuffer, 0x1A - aFirstAddress, int16_t(mPitchGyro->GetAngle() * 100), 2); // Pitch
	PutTheValue(aBuffer, 0x18 - aFirstAddress, int16_t(mRollGyro->GetAngle() * 100), 2); // Roll
	PutTheValue(aBuffer, 0x1C - aFirstAddress, uint16_t(0 * 100), 2); // Heading
	PutTheValue(aBuffer, 0x32 - aFirstAddress, int16_t(0 * 100), 2); // Temperature
	PutTheValue(aBuffer, 0x24 - aFirstAddress, int16_t(0 * 1000), 2); // Linear Accel X
	PutTheValue(aBuffer, 0x26 - aFirstAddress, int16_t(0 * 1000), 2); // Linear Accel Y
	PutTheValue(aBuffer, 0x28 - aFirstAddress, int16_t(0 * 1000), 2); // Linear Accel Z

}

File: snobot_sim_java/java/snobot_sim_java/com/snobot/simulator/simulator_components/navx/NavxSimulator.java
package com.snobot.simulator.simulator_components.navx;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

import com.snobot.simulator.SensorActuatorRegistry;
import com.snobot.simulator.simulator_components.gyro.GyroWrapper;
import com.snobot.simulator.simulator_components.gyro.GyroWrapper.AngleSetterHelper;

public class NavxSimulator
{

    protected static int TIMESTAMP = 1;

    protected final GyroWrapper mYawGyro;
    protected final GyroWrapper mPitchGyro;
    protected final GyroWrapper mRollGyro;
    protected final int mNativePort;

    public NavxSimulator(int aNativePort, int aSimulatorOffset)
    {
        mYawGyro = new GyroWrapper("NavX Pitch", NULL_ANGLE_SETTER);
        mPitchGyro = new GyroWrapper("NavX Yaw", NULL_ANGLE_SETTER);
        mRollGyro = new GyroWrapper("NavX Roll", NULL_ANGLE_SETTER);

        mNativePort = aNativePort;

        int basePort = aSimulatorOffset + aNativePort * 3;

        SensorActuatorRegistry.get().register(mYawGyro, basePort + 0);
        SensorActuatorRegistry.get().register(mPitchGyro, basePort + 1);
        SensorActuatorRegistry.get().register(mRollGyro, basePort + 2);
    }

    protected ByteBuffer createConfigBuffer()
    {
        ByteBuffer output = ByteBuffer.allocateDirect(17);

        output.put((byte) 1);
        output.put((byte) 2);
        output.put((byte) 3);
        output.put((byte) 4);
        output.put((byte) 5);
        output.putShort((short) 0x0607);
        output.put((byte) 6);
        output.put((byte) 7);
        output.put((byte) 10);
        output.putShort((short) 0xDEAD);
        output.putInt(0);
        output.putShort((short) 0xBEEF);

        return output;
    }

    protected ByteBuffer createDataBuffer(int aFirstAddress)
    {
        ByteBuffer output = ByteBuffer.allocateDirect(86 - aFirstAddress);

        output.order(ByteOrder.LITTLE_ENDIAN);
        output.putInt(0x12 - aFirstAddress, TIMESTAMP);
        ++TIMESTAMP;

        output.put(0x08 - aFirstAddress, (byte) 1); // Op Status
        output.put(0x0A - aFirstAddress, (byte) 2); // Self Test
        output.put(0x09 - aFirstAddress, (byte) 3); // Cal Status
        output.put(0x10 - aFirstAddress, (byte) 4); // Sensor Status

        output.putShort(0x16 - aFirstAddress, (short) (boundBetweenNeg180Pos180(mYawGyro.getAngle()) * 100)); // Yaw
        output.putShort(0x1A - aFirstAddress, (short) (boundBetweenNeg180Pos180(mPitchGyro.getAngle()) * 100)); // Pitch
        output.putShort(0x18 - aFirstAddress, (short) (boundBetweenNeg180Pos180(mRollGyro.getAngle()) * 100)); // Roll
        output.putShort(0x1C - aFirstAddress, (short) (0 * 100)); // Heading
        output.putShort(0x32 - aFirstAddress, (short) (0 * 100)); // Temperature
        output.putShort(0x24 - aFirstAddress, (short) (0 * 1000)); // Linear
                                                                   // Accel X
        output.putShort(0x26 - aFirstAddress, (short) (0 * 1000)); // Linear
                                                                   // Accel Y
        output.putShort(0x28 - aFirstAddress, (short) (0 * 1000)); // Linear
                                                                   // Accel Z

        return output;
    }

    private double boundBetweenNeg180Pos180(double input)
    {
        double output = input;
        while (output > 180)
        {
            output -= 360;
        }
        while (output < -180)
        {
            output += 360;
        }

        return output;
    }

    protected static final GyroWrapper.AngleSetterHelper NULL_ANGLE_SETTER = new AngleSetterHelper()
    {

        @Override
        public void updateAngle(double aAngle)
        {
            // Nothing to do, handled by read/write transactions
        }
    };
}
